sync/optimistic.md

```diff
@@ -0,0 +1,340 @@
+blocks without verifying the execution payloads. This partial sync is called an
+*optimistic sync*.
+
+## Constants
+
+|Name|Value|Unit
+|---|---|---|
+|`SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY`| `128` | slots
+
+*Note: the `SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY` must be user-configurable. See
+[Fork Choice Poisoning](#fork-choice-poisoning).*
+
+## Helpers
+
+Let `head: BeaconBlock` be the result of calling of the fork choice
+algorithm at the time of block production. Let `head_block_root: Root` be the
+root of that block.
+
+Let `blocks: Dict[Root, BeaconBlock]` and `block_states: Dict[Root,
+BeaconState]` be the blocks (and accompanying states) that have been verified
+either completely or optimistically.
+
+Let `optimistic_roots: Set[Root]` be the set of `hash_tree_root(block)` for all
+optimistically imported blocks which have only received a `SYNCING` designation
+from an execution engine (i.e., they are not known to be `INVALID` or `VALID`).
+
+```python
+@dataclass
+class Store(object):
+    optimistic_roots: Set[Root]
+    head_block_root: Root
+    blocks: Dict[Root, BeaconBlock]
+    block_states: Dict[Root, BeaconState]
+```
+
+```python
+def is_optimistic(store: Store, block: BeaconBlock) -> bool:
+    return hash_tree_root(block) in store.optimistic_roots
+```
+
+```python
+def latest_verified_ancestor(store: Store, block: BeaconBlock) -> BeaconBlock:
+    # It is assumed that the `block` parameter is never an INVALID block.
+    while True:
+        if not is_optimistic(store, block) or block.parent_root == Root():
+            return block
+        block = store.blocks[block.parent_root]
+```
+
+```python
+def is_execution_block(block: BeaconBlock) -> BeaconBlock:
+    return block.body.execution_payload != ExecutionPayload()
+```
+
+```python
+def should_optimistically_import_block(store: Store, current_slot: Slot, block: BeaconBlock) -> bool:
+    justified_root = store.block_states[store.head_block_root].current_justified_checkpoint.root
+    justifed_is_verified = is_execution_block(store.blocks[justified_root])
+    block_is_deep = block.slot + SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY <= current_slot
+    return justified_is_verified or block_is_deep
+```
+
+Let only a node which returns `is_optimistic(store, head) is True` be an *optimistic
+node*. Let only a validator on an optimistic node be an *optimistic validator*.
+
+When this specification only defines behaviour for an optimistic
+node/validator, but *not* for the non-optimistic case, assume default
+behaviours without regard for optimistic sync.
+
+## Mechanisms
+
+### When to optimistically import blocks
+
+A block MAY be optimistically imported when
+`should_optimistically_import_block(store, current_slot, block)` returns
+`True`. This ensures that blocks are only optimistically imported if either:
+
+1. The justified checkpoint has execution enabled.
+1. The current slot (as per the system clock) is at least
```

Comment:
Suggest defining `current_slot` wrt wall clock time in the helper exposition. Otherwise, it isn't clear when just reading `should_optimistically_import_block` what it means in that context

Reply:
Nice one, addressed in bfe41725